src/services/codefixes/convertToAsyncFunction.ts

```diff
@@ -0,0 +1,471 @@
+                        lastDotThen.set(getNodeId(child).toString(), false);
+                    }
+                    else if (isCallExpression(child) || isIdentifier(child) && nodeType && checker.getPromisedTypeOfPromise(nodeType)) {
+
+                        // true - there is a following .then() in the callback chain
+                        lastDotThen.set(getNodeId(child).toString(), true);
+                    }
+                });
+            }
+            else {
+                forEachChild(node, visit);
+            }
+        });
+
+        return lastDotThen;
+    }
+
+    function isFunctionRef(node: Node): boolean {
+        const callExpr = climbPastPropertyAccess(node);
+        return !isCallExpression(callExpr) || callExpr.expression !== node;
+    }
+
+    function declaredInFile(symbol: Symbol, sourceFile: SourceFile): boolean {
+        return symbol.valueDeclaration && symbol.valueDeclaration.getSourceFile() === sourceFile;
+    }
+
+    // varNamesMap holds all of the variables in original source code. synthNamesMap holds all of the variables created by the refactor
+    function renameCollidingVarNames(nodeToRename: FunctionLikeDeclaration, checker: TypeChecker, synthNamesMap: Map<SynthIdentifier>, context: CodeFixContextBase, lastDotThenMap: Map<boolean>, originalType: Map<Type>): FunctionLikeDeclaration {
+        const allVarNames: [Identifier, Symbol][] = [];
+
+        forEachChild(nodeToRename, function visit(node: Node) {
+            const symbol = checker.getSymbolAtLocation(node);
+            const isDefinedInFile = symbol ? declaredInFile(symbol, context.sourceFile) : undefined;
+
+            if (isIdentifier(node) && symbol && isDefinedInFile) {
+                const type = checker.getTypeAtLocation(node);
+
+                // if the identifier refers to a function
+                if (type && type.getCallSignatures().length > 0 && isFunctionRef(node)) {
+                    if (type.getCallSignatures()[0].parameters.length && !synthNamesMap.get(getSymbolId(symbol).toString())) {
+                        // add the new synthesized variable for the declaration (ex. blob in let blob = res(arg))
+                        const synthName = getNewNameIfConflict(createIdentifier(type.getCallSignatures()[0].parameters[0].name), allVarNames);
+                        synthNamesMap.set(getSymbolId(symbol).toString(), synthName);
+                        allVarNames.push([synthName.identifier, symbol]);
+                    }
+                }
+                else if (node.parent && isParameter(node.parent) || isVariableDeclaration(node.parent)) {
+                    const newName = getNewNameIfConflict(node, allVarNames);
+                    let setName = false;
+
+                    for (const ident of allVarNames) {
+                        if (ident[0].text === node.text && ident[1] !== symbol) {
+                            synthNamesMap.set(getSymbolId(symbol).toString(), newName);

```

Comment:
The body of the `if` doesn't seem to depend on the loop variable.  Break after it's executed?

Reply:
Looks like you could even return and eliminate the flag variable.